arbnode/transaction_streamer.go

```diff
@@ -512,6 +526,40 @@ func (s *TransactionStreamer) SequenceTransactions(header *arbos.L1IncomingMessa
 		delayedMessagesRead = lastMsg.DelayedMessagesRead
 	}
 
+	if nonceCacheSize > 0 {
+		oldPreTxFilter := hooks.PreTxFilter
+		oldPostTxFilter := hooks.PostTxFilter
+		hooks.PreTxFilter = func(config *params.ChainConfig, header *types.Header, statedb *state.StateDB, arbos *arbosState.ArbosState, tx *types.Transaction, sender common.Address) error {
+			stateNonce, ok := s.nonceCache.Get(sender)
+			if !ok {
+				stateNonce = statedb.GetNonce(sender)
+				s.nonceCache.Add(sender, stateNonce)
+			}
+			err := MakeNonceError(sender, tx.Nonce(), stateNonce)
+			if err != nil {
+				return err
+			}
+			return oldPreTxFilter(config, header, statedb, arbos, tx, sender)
```

Comment:
Wouldn't this add an extra prefilter to the chain every single time `SequenceTransactions` is called?

Reply:
Hmm.. Yeah, what I'd really like is a rust generic which would be able to do this zero-cost, but without that option I'll refactor this out. I initially had the nonce cache living in the sequencer itself, but because of the createBlocks clear and the clear on sequencing error I think it needs to live in the transaction streamer. That said, I can have the sequencer pre-check hook call back into the tx streamer, which is what I'll do.

Reply:
I've fixed this by having sequencer call into the transaction streamer explicitly in its filters